src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+
+        sprintf(buf, "%02x", b);
+        name.push_back(buf[0]);
+        name.push_back(buf[1]);
+    }
+
+    assert(name.length() == sizeof(extaddr) * 2);
+
+    otbrLogDebug("Using instance name %s", name.c_str());
+    return name;
+}
+
+void TrelDnssd::DoRegisterTrelService(void)
+{
+    Mdns::Publisher::SubTypeList subTypeList;
+
+    assert(mRegisterInfo.IsValid());
+    assert(!mRegisterInfo.IsRegistered());
+    assert(mTrelNetifIndex > 0);
+
+    mRegisterInfo.mInstanceName = GetTrelInstanceName();
+    mPublisher.PublishService("", mRegisterInfo.mPort, mRegisterInfo.mInstanceName, kTrelServiceName, subTypeList,
+                              mRegisterInfo.mTxtEntries);
+}
+
+void TrelDnssd::DoRemoveTrelService(void)
+{
+    assert(mRegisterInfo.IsValid());
+    assert(mRegisterInfo.IsRegistered());
+
+    mPublisher.UnpublishService(mRegisterInfo.mInstanceName, kTrelServiceName);
+    mRegisterInfo.mInstanceName = "";
+}
+
+void TrelDnssd::OnTrelServiceAdded(const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    std::string        instanceName = StringUtils::ToLowercase(aInstanceInfo.mName);
+    otPlatTrelPeerInfo peerInfo;
+
+    // Remove any existing TREL service instance before adding
+    RemoveTrelService(instanceName);
+
+    otbrLogDebug("Peer discovered: %s hostname %s addresses %zu port %d priority %d "
+                 "weight %d",
+                 aInstanceInfo.mName.c_str(), aInstanceInfo.mHostName.c_str(), aInstanceInfo.mAddresses.size(),
+                 aInstanceInfo.mPort, aInstanceInfo.mPriority, aInstanceInfo.mWeight);
+
+    for (auto &addr : aInstanceInfo.mAddresses)
+    {
+        otbrLogDebug("Peer address: %s", addr.ToString().c_str());
+    }
+
+    if (aInstanceInfo.mAddresses.empty())
+    {
+        otbrLogWarning("Peer %s does not have any IPv6 address, ignored", aInstanceInfo.mName.c_str());
+        ExitNow();
+    }
+
+    {
+        Peer peer(aInstanceInfo.mTxtData, peerInfo.mSockAddr);
+
+        VerifyOrExit(peer.mValid, otbrLogWarning("Peer %s is invalid", aInstanceInfo.mName.c_str()));
+
+        peerInfo.mRemoved = false;
+        memcpy(&peerInfo.mSockAddr.mAddress, &aInstanceInfo.mAddresses[0], sizeof(peerInfo.mSockAddr.mAddress));
+        peerInfo.mSockAddr.mPort = aInstanceInfo.mPort;
+        peerInfo.mTxtData        = aInstanceInfo.mTxtData.data();
+        peerInfo.mTxtLength      = aInstanceInfo.mTxtData.size();
+
+        otPlatTrelHandleDiscoveredPeerInfo(mNcp.GetInstance(), &peerInfo);
+
+        mPeers.emplace(instanceName, peer);
+        CheckPeersNumLimit();
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RemoveTrelService(const std::string &aInstanceName)
+{
+    std::string instanceName = StringUtils::ToLowercase(aInstanceName);
+    auto        it           = mPeers.find(instanceName);
+
+    VerifyOrExit(it != mPeers.end());
+
+    otbrLogDebug("Peer removed: %s", instanceName.c_str());
+
+    if (CountDuplicatePeers(it->second) == 0)
```

Comment:
Consider the following case:
1. Peer A advertises TREL service with instance name "A1"
1. Peer A rebooted without having a chance to remove TREL service
1. Peer A advertises TREL service with instance name "A2"

So, on Peer B, the "A1" instance will expire and eventually trigger an instance removal. So, Peer B should not trigger actual peer removing because "A2" is still alive. 

 Another way of handling it may be Peer B removing instance "A1" automatically when it discovered "A2". But, the current implementation didn't go this way.

Reply:
Thanks. I see. Yes I think this would work.